### ISSUE TITLE
Refactor pgx and postgres db drivers

### ISSUE THREAD

**Issue by dhui** [golang-migrate/migrate#549]
**Is your feature request related to a problem? Please describe.**
The `pgx` and `postgres` drivers share a lot of code since the `pgx` driver was copied from the `postgres` driver. This makes development more onerous since a fix will generally affect both drivers and needs to be applied to both.

**Describe the solution you'd like**
Factor out common code and tests to an `internal/postgres` driver that's imported by the `pgx` and `postgres` drivers.

**Describe alternatives you've considered**
1. Leave as is and wait for postgres driver deprecation

**Additional context**
N/A

**Comment by oldgalileo** [comment]
Understanding I've never interacted with the `migrate` team, I'd like to advocate for speedy deprecation of the postgres driver built on `lib/pq` rather than adding any additional complexity through a common postgres driver base.

Given that `lib/pq` itself is advocating for `pgx` and with the fact that `pgx` seems [faster than `lib/pq` across the board](https://devandchill.com/posts/2020/05/go-lib/pq-or-pgx-which-performs-better/), it just doesn't seem worth the time.

**Comment by dhui** [comment]
@oldgalileo Are you suggesting that the migrate pgx driver should [use the pgx interface instead of the `database/sql` one](https://github.com/jackc/pgx#choosing-between-the-pgx-and-databasesql-interfaces)?
That is an option. I'm wondering if there's an internal interface we can define that'll work with our usage of both `*sql.DB` and `*pgx.Conn` so we can continue to use the same logic to run postgres migrations between both drivers.
However, performance isn't a priority for this project. Correctness, maintainability, and backwards compatibility are more important for this project, so if an interface working with both `*sql.DB` and `*pgx.Conn` can't be defined, we'll need to continue using the pgx `database/sql` interface.

Speaking of backwards compatibility, to be clear, support for `lib/pq` will remain for a while to support teams still using it.

**Comment by oldgalileo** [comment]
@dhui Yes to your first question of using `pgx` 'natively' over the `database/sql` interface. Although as I stopped for a second and thought about how to respond here, I realized the bigger issue is simply the absence of a `pgx`-bundled `migrate` release. Having to include `lib/pq` and its dependencies in addition to `pgx` and ITS dependencies adds unnecessary bloat to a project. 

Given that `pgx` has gained a lot of traction over `lib/pq`, I'd imagine many apps are fraught with added dependencies. Perhaps an immediate solution to _that_ issue is to simply include a brief message specifically in the `database/pgx/README.md` on [importing pseudo versions](https://golang.org/ref/mod#versions), like so:

```
Warning: Using pseudo-versions and not adhering to the releases or their candidates should be done at your own risk. These imports are not released after rigorous testing through the CI pipeline and there may be dormant issues not-yet caught.

require github.com/golang-migrate/migrate/v4 v4.0.0-20210505171241-ffdcb52998c9
```

While there is clearly merit in deciding how best to approach a unified interface for `*sql.DB` and `*pgx.Conn`, I'd like to move the goalposts here (perhaps I'll open a PR if deemed fit).

**Comment by dhui** [comment]
@oldgalileo 
Release are [blocked](https://github.com/golang-migrate/migrate/issues/252#issuecomment-767399329) until CI supports releases again. This [PR](https://github.com/golang-migrate/migrate/pull/521) is the most promising but I haven't had time to test it.
There's a [separate issue](https://github.com/golang-migrate/migrate/issues/174) around splitting up the drivers.

I'm not sure how pseudo versions are relevant to our discussion.

> Given that `pgx` has gained a lot of traction over `lib/pq`

How are you measuring this? Looking at their [GitHub star history](https://star-history.t9t.io/#lib/pq&jackc/pgx), `pq` is still more popular than `pgx` but `pgx` is quickly gaining.

> I'd like to move the goalposts here (perhaps I'll open a PR if deemed fit).

What's your proposal?

**Comment by craigpastro** [comment]
It would also be really nice if we could still start the connection string with `postgres://` and let the import handle which driver is used under the hood.

**Comment by dhui** [comment]
@craigpastro https://github.com/golang-migrate/migrate/blob/master/database/pgx/v5/

**Comment by ptrus** [comment]
`pgx` itself doesn't support the "pgxv5:" connection string:
https://github.com/jackc/pgx/blob/1b68b5970e8329ad3035b32a97bc78e38207951e/pgconn/config.go#L232C1-L232C102

So it's slightly inconvenient not being able to use the same connection string for configuring a plain `pgx` client and setting up the `pgx`-backed migration client.
